rig/machine_control/scp_connection.py

```diff
@@ -79,65 +125,167 @@ def send_scp(self, buffer_size, x, y, p, cmd, arg1=0, arg2=0, arg3=0,
+                        seq = next(self.seq)
+
+                    # Construct the packet that we'll be sending
+                    packet = SCPPacket(
+                        reply_expected=True, tag=0xff, dest_port=0,
+                        dest_cpu=args.p, src_port=7, src_cpu=31,
+                        dest_x=args.x, dest_y=args.y, src_x=0, src_y=0,
+                        cmd_rc=args.cmd, seq=seq,
+                        arg1=args.arg1, arg2=args.arg2, arg3=args.arg3,
+                        data=args.data
                     )
-                )
 
-            if scp.seq == self._seq:
-                # The packet is the acknowledgement.  Increment the
-                # sequence indicator and return the packet.
-                self._seq ^= 1
-                return scp
+                    # Create a reference to this packet so that we know we're
+                    # expecting a response for it and can retransmit it if
+                    # necessary.
+                    outstanding_packets[seq] = TransmittedPacket(
+                        args.callback, packet.bytestring, args.expected_args,
+                        args.timeout
+                    )
 
-        # The packet we transmitted wasn't acknowledged.
-        raise TimeoutError(
-            "Exceeded {} tries when trying to transmit packet.".format(
-                self.n_tries)
-        )
+                    # Actually send the packet
+                    self.sock.send(packet.bytestring)
 
-    def read(self, buffer_size, x, y, p, address, length_bytes):
+            # Listen on the socket for an acknowledgement packet, there not be
+            # one.
+            try:
+                ack = self.sock.recv(receive_length)
+            except IOError:
+                # There wasn't a returned packet, we may spend quite some time
+                # here.
+                ack = None
+
+            # Process the received packet (if there is one)
+            if ack is not None:
+                # Extract the sequence number from the bytestring, iff possible
+                rc, seq = struct.unpack_from("<2H", ack,
+                                             consts.SDP_HEADER_LENGTH + 2)
+
+                # If the code is an error then we respond immediately
+                if rc in self.error_codes:
+                    raise self.error_codes[rc]
+
+                # Look up the sequence index of packet in the list of
+                # outstanding packets.  We may have already processed a
+                # response for this packet (indicating that the response was
+                # delayed and we retransmitted the initial message) in which
+                # case we can silently ignore the returned packet.
+                # XXX: There is a danger that a response was so delayed that we
+                # already reused the seq number... this is probably
+                # sufficiently unlikely that there is no problem.
```

Comment:
Worth noting that the sequence values go up to 65535, so the likelihood of this occurring *is* small.